bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java

```diff
@@ -827,4 +829,25 @@ private void resetSize(int index) {
             size = index + 1;
         }
     }
+
+    private void validateTupleSizeAndInherentType() {
+        int numOfMandatoryTypes = this.tupleType.getTupleTypes().size();
+        // Check if tuple has at least one more members than the number of mandatory type to compensate the
+        // removed member
+        if (numOfMandatoryTypes >= this.getLength()) {
+            throw ErrorHelper.getRuntimeException(
+                    getModulePrefixedReason(ARRAY_LANG_LIB, SIZE_MISMATCH_ERROR_IDENTIFIER),
+                    ErrorCodes.INVALID_MEMBER_SIZE, numOfMandatoryTypes + 1, this.getLength());
+        }
+        // Check if value belonging to i th type can be assigned to i-1 th type (Checking done by value, not type)
+        for (int i = 1; i <= numOfMandatoryTypes; i++) {
+            if (!TypeChecker.checkIsType(this.getRefValue(i), this.tupleType.getTupleTypes().get(i - 1))) {
+                throw ErrorHelper.getRuntimeException(
+                        getModulePrefixedReason(ARRAY_LANG_LIB, INHERENT_TYPE_VIOLATION_ERROR_IDENTIFIER),
+                        ErrorCodes.INCOMPATIBLE_TYPE, this.tupleType.getTupleTypes().get(i - 1),
+                        (i == numOfMandatoryTypes) ?
+                                this.tupleType.getRestType() : this.tupleType.getTupleTypes().get(i));
```

Comment:
Why do we need this check? Don't we check this at compile time within the types of tuples? Can you mention a test that fails due to this check?

Reply:
```
[int, int, string...] tuple1 = [0, 1, "hello"];
[anydata, anydata, anydata...] tuple2 = tuple1;
var x = tuple2.shift();
```
The above code compiles, but it should panic at runtime to preserve the inherent type of member 2 when we try to assign it the value "hello" while it expects an `int`.

Reply:
This is working as expected in update 9 as well, isn't it?

Reply:
Currently, the shift operation for tuples is not supported by the runtime. Therefore the runtime checks if the object performing the shift is an array and if not, throws an `OperationNotSupported` exception. The changes proposed in this PR allows the use of shift operation on tuples but this also allows the risk of running into mutations that violate the inherent type of the tuple (as shown in the previous comment). The check in question is added to make sure that never happens.

Reply:
Okay, got it. We introduced this after update 9, right?